### PR TITLE
handle the case smoother when all endpoints are fading in

### PR DESCRIPTION
**Pull request by aryszka**
use weighted shifting instead of shifting to oldest when there are no available stable endpoints

### CODE REVIEW

File: loadbalancer/algorithm.go [hunk 1]
```diff
@@ -101,14 +117,15 @@ func withFadeIn(rnd *rand.Rand, ctx *routing.LBContext, w []int, choice int) rou
...
 type roundRobin struct {
-	mx         sync.Mutex
-	index      int
-	rnd        *rand.Rand
-	fadeInCalc []int
+	mx               sync.Mutex
+	index            int
+	rnd              *rand.Rand
+	fadeInCalcInts   []int
+	fadeInCalcFloats []float64
```
Comment (szuecs) [inline loadbalancer/algorithm.go:128]: fadeInCalcInts and fadeInCalcFloats are not obvious names. 
fadeInCalcInts is stored in notFadingIndexes, which is much more clear IMO. 
And fadeInCalcFloats seem to be fadingWeights or fadingPercentage.

wdyt?

File: loadbalancer/algorithm.go [hunk 2]
```diff
@@ -55,20 +55,36 @@ func fadeIn(now time.Time, duration time.Duration, exponent float64, detected ti
...
-	for _, epi := range ep[1:] {
-		if epi.Detected.Before(oldest.Detected) {
-			oldest = epi
+	for _, epi := range ep {
+		wi := fadeIn(now, rt.LBFadeInDuration, rt.LBFadeInExponent, epi.Detected)
+		sum += wi
+		weightSums = append(weightSums, sum)
+	}
+
+	var choice routing.LBEndpoint
```
Comment (AlexanderYastrebov) [inline loadbalancer/algorithm.go:69]: Wouldn't `choice := ep[len(ep)-1]` eliminate `if i == len(weightSums)-1` in the loop?

**Comment by AlexanderYastrebov** [comment]
I did a synthetic test for fresh start case:
```
$ tail -n 10 /etc/hosts
127.0.0.1 h0.test
127.0.0.1 h1.test
127.0.0.1 h2.test
127.0.0.1 h3.test
127.0.0.1 h4.test
127.0.0.1 h5.test
127.0.0.1 h6.test
127.0.0.1 h7.test
127.0.0.1 h8.test
127.0.0.1 h9.test

$ cat ~/tmp/algo.eskip 
r0: Host("h0.test") -> status(200) -> <shunt>;
r1: Host("h1.test") -> status(210) -> <shunt>;
r2: Host("h2.test") -> status(220) -> <shunt>;
r3: Host("h3.test") -> status(230) -> <shunt>;
r4: Host("h4.test") -> status(240) -> <shunt>;
r5: Host("h5.test") -> status(250) -> <shunt>;
r6: Host("h6.test") -> status(260) -> <shunt>;
r7: Host("h7.test") -> status(270) -> <shunt>;
r8: Host("h8.test") -> status(280) -> <shunt>;
r9: Host("h9.test") -> status(290) -> <shunt>;

test1: Path("/test") -> setPath("/") -> fadeIn("3m") -> 
<
    roundRobin, // algorithm
    "http://h0.test:9090",
    "http://h1.test:9090",
    "http://h2.test:9090",
    "http://h3.test:9090",
    "http://h4.test:9090",
    "http://h5.test:9090",
    "http://h6.test:9090",
    "http://h7.test:9090",
    "http://h8.test:9090",
    "http://h9.test:9090"
>;

bin/skipper -routes-file ~/tmp/algo.eskip -access-log-disabled
```

On master branch (just after skipper start) there is:
```
$ echo "GET http://localhost:9090/test" | vegeta attack -duration=20s -rate=100/s | vegeta report;
Requests      [total, rate, throughput]         2000, 100.05, 100.04
Duration      [total, attack, wait]             19.991s, 19.99s, 1.295ms
Latencies     [min, mean, 50, 90, 95, 99, max]  290.588µs, 1.299ms, 1.33ms, 1.591ms, 1.735ms, 2.884ms, 6.153ms
Bytes In      [total, mean]                     0, 0.00
Bytes Out     [total, mean]                     0, 0.00
Success       [ratio]                           100.00%
Status Codes  [code:count]                      200:1859  210:16  220:11  230:15  240:14  250:18  260:17  270:13  280:23  290:14  
Error Set:
```
a clear skew towards first endpoint.

On this branch it looks fine:
```
$ echo "GET http://localhost:9090/test" | vegeta attack -duration=20s -rate=100/s | vegeta report;
Requests      [total, rate, throughput]         2000, 100.05, 100.04
Duration      [total, attack, wait]             19.991s, 19.99s, 1.512ms
Latencies     [min, mean, 50, 90, 95, 99, max]  312.756µs, 1.473ms, 1.439ms, 1.706ms, 1.9ms, 2.877ms, 6.977ms
Bytes In      [total, mean]                     0, 0.00
Bytes Out     [total, mean]                     0, 0.00
Success       [ratio]                           100.00%
Status Codes  [code:count]                      200:184  210:194  220:217  230:207  240:224  250:181  260:192  270:215  280:195  290:191  
Error Set:
```

File: loadbalancer/algorithm.go (hunk 1 above)
Comment (aryszka) [reply loadbalancer/algorithm.go]: sounds good to me, thx! Might happen that earlier the ints were used for different purposes as worker slices, but now there is only one.

File: loadbalancer/algorithm.go (hunk 2 above)
Comment (aryszka) [reply loadbalancer/algorithm.go]: true, thx!

**Comment by szuecs** [comment]
👍

**Comment by aryszka** [comment]
:+1:

**Comment by szuecs** [comment]
This change created a fix but tests that start with no endpoints and create multiple new endpoints were flaky because the tests did not change to make sure that we do not fadein but have an even distribution.

see also https://github.com/zalando/skipper/pull/3786/files#diff-ca886c0604353f1f9ef28c5d0dd9bc41ed172442c5dfe26c17c3aed440a180d5R493-R507

**Pull request merged**